### PR TITLE
[Spec Decode] Introduce DraftModelRunner

### PR DESCRIPTION
**Pull request by comaniac**
This PR introduces a DraftModelRunner for speculative decoding. The unique feature of DraftModelRunner is allowing multiple steps in one `execute_model()`. In this way, the following features can be implemented inside this runner in parallel without affecting non-speculative decoding use cases:

* Bonus token (cc @sroy745)
* Async prepare_inputs for the next step (cc @alexm-neuralmagic)
* Keep prepare_input on GPU (cc @alexm-neuralmagic, @rkooo567)

Also cc @cadedaniel

### CODE REVIEW

**Comment by comaniac** [comment]
> Who is shepherding this PR?

@cadedaniel should be the one, but I'd prefer to rebase after #5408 (@youkaichao can we get that PR merged by today?)

**Comment by cadedaniel** [comment]
Ok I will review today.

File: vllm/spec_decode/multi_step_worker.py
Comment (sroy745) [inline vllm/spec_decode/multi_step_worker.py:70]: As per the offline discussions on slack this is my understanding of how we will plugin the bonus token logic here. Please confirm if my understanding is correct

```
expanded_request, original_seq_indices = self.expand_batch(
    copied_execute_model_req, seq_ids_with_bonus_tokens)
model_outputs: List[SamplerOutput] = self.execute_model(
            execute_model_req=expanded_request)
filtered_model_output = filter_output(model_outputs, original_seq_indices)
.....
.....

def expand_batch(
    self, copied_execute_model_req, seq_ids_with_bonus_tokens):
"""
Takes the original request and expands it for sequences which have bonus tokens. Returns
a new request with the expanded batch and also the indices of the original sequences.
"""
....
....


def filter_output(
    self, model_outputs, original_seq_indices):
"""
Iterate over model_outputs and filters the output of each step. 
The filtering retains the output corresponding to indices specified in original_seq_indices.
Returns the filtered output.

"""
.....
.....

```

**Review by sroy745 — COMMENTED** [review]
Thanks for the PR. One clarifying question about how the bonus token logic will fit in the new flow.

File: vllm/worker/worker.py
Comment (comaniac) [reply vllm/worker/worker.py:265]: Hmm actually we don't. If it's not the driver worker than it won't reach here. Thanks for the good catch and I'll remove it.

File: vllm/spec_decode/multi_step_worker.py
Comment (comaniac) [reply vllm/spec_decode/multi_step_worker.py:70]: Yes this is also my understanding. One miner point to clarify is we will put the num_steps:

```python
model_outputs: List[SamplerOutput] = self.execute_model(
    execute_model_req=expanded_request, num_steps=num_steps
)
```

File: vllm/worker/worker.py
Comment (cadedaniel) [inline vllm/worker/worker.py:302]: Seems the model runner should return List, then the Worker returns the model runner output.

File: vllm/spec_decode/draft_model_runner.py
Comment (cadedaniel) [inline vllm/spec_decode/draft_model_runner.py:15]: can we add a fat docstring for this class, explaining the motivation and high-level plans?

**Review by cadedaniel — APPROVED** [review]
LGTM, nice!
* can we get @sroy745 's approval before merge? want to make sure bonus token work is compatible
* +1 to waiting for https://github.com/vllm-project/vllm/pull/5408

**Review by sroy745 — APPROVED** [review]
Thanks for confirming. LGTM.

**Comment by comaniac** [comment]
Thanks for the review. Summarizing the action items:
1. Rebase after #5408.
2. Docstring.
3. Model runner returns a list.

**Comment by comaniac** [comment]
Rebased after #5408 is merged. Note that since now `model_runner.execute_model` no longer takes `seq_group_metadata_list`, we cannot directly call `.prepare_model_input` inside `model_runner.execute_model` but have to in-place update `model_input` as planned. In order to keep this PR lightweight, I currently use a temporary solution that caches `seq_group_metadata_list`, and this should be removed once @alexm-neuralmagic updated `update_model_input`.

**Comment by comaniac** [comment]
Hmm somehow return a list of SamplerOutput from `model_runner.execute_mode` causes hangin with TP>1. I need to investigate the root cause...

**Comment by comaniac** [comment]
Also cc @rkooo567 @LiuXiaoxuanPKU for a second view

**Pull request merged**
